core/eolearn/core/utils/parsing.py

```diff
@@ -13,15 +13,32 @@
 """
 from __future__ import annotations
 
+import sys
 from itertools import repeat
-from typing import TYPE_CHECKING, Iterable, List, Optional, Sequence, Tuple, Union, cast
+from typing import TYPE_CHECKING, Dict, Iterable, List, Optional, Sequence, Tuple, Union, cast
 
 from ..constants import FeatureType
+from .types import EllipsisType
+
+if sys.version_info < (3, 8):
+    from typing_extensions import Literal
+else:
+    from typing import Literal  # pylint: disable=ungrouped-imports
 
 if TYPE_CHECKING:
     from ..eodata import EOPatch
 
 
+SingleFeatureSpec = Union[
+    Literal[FeatureType.TIMESTAMP, FeatureType.BBOX], Tuple[FeatureType, str], Tuple[FeatureType, str, str]
+]
+SequenceFeatureSpec = Sequence[Union[SingleFeatureSpec, Tuple[FeatureType, EllipsisType]]]
+
+DictFeatureSpec = Dict[FeatureType, Union[None, EllipsisType, Iterable[Union[str, Tuple[str, str]]]]]
```

Comment:
I assume the `Dict[FeatureType, Union[None]]` part is for the `TIMESTAMP` and `BBOX`?

If yes, then perhaps would it be more explicit to do:

```python
BaseFeatureSpec = Literal[FeatureType.TIMESTAMP, FeatureType.BBOX]
DictFeatureSpec = Union[Dict[BaseFeatureSpec, None], Dict[FeatureType, Union[EllipsisType, Iterable[Union[str, Tuple[str, str]]]]]]
```

where `BaseFeatureSpec` could be used also above.

Reply:
what you wrote wont work for `{FeatureType.BBOX: None, FeatureType.DATA: ...}`

Reply:
this could perhaps be solved with typed dictionaries, but I'm afraid of adding even more types to this monstrosity. I'd prefer to greatly reduce the parser, but open source is a curse in that regard.

Reply:
OK, thanks for pointing out that use case. Resolving.